File: 2946-matrix-similarity-after-cyclic-shifts/2946-matrix-similarity-after-cyclic-shifts.cpp
class Solution {
public:
    bool areSimilar(vector<vector<int>>& mat, int k) {
        int n = mat.size(), m = mat[0].size();
        k %= m;
        
        for (int i = 0; i < n; i++) {
            vector<int> temp(m);
            for (int j = 0; j < m; j++) {
                if (i % 2 == 0)
                    temp[j] = mat[i][(j + k) % m];
                else
                    temp[j] = mat[i][(j - k + m) % m];
            }
            if (temp != mat[i]) return false;
        }
        return true;
    }
};

File: 2946-matrix-similarity-after-cyclic-shifts/2946-matrix-similarity-after-cyclic-shifts.dart
class Solution {
  bool areSimilar(List<List<int>> mat, int k) {
    int n = mat.length, m = mat[0].length;
    k %= m;

    for (int i = 0; i < n; i++) {
      List<int> temp = List.filled(m, 0);
      for (int j = 0; j < m; j++) {
        if (i % 2 == 0) {
          temp[j] = mat[i][(j + k) % m];
        } else {
          temp[j] = mat[i][(j - k % m + m) % m];
        }
      }
      for (int j = 0; j < m; j++) {
        if (temp[j] != mat[i][j]) return false;
      }
    }
    return true;
  }
}